libgadget/fof.c

```diff
@@ -501,16 +519,16 @@ fof_primary_ngbiter(TreeWalkQueryFOF * I,
     else /* mode is 1, target is a ghost */
     {
         int head = HEAD(other, FOF_PRIMARY_GET_PRIV(tw)->Head);
-        struct SpinLocks * spin = FOF_PRIMARY_GET_PRIV(tw)->spin;
-//        printf("locking %d by %d in ngbiter\n", other, omp_get_thread_num());
-        lock_spinlock(head, spin);
-        if(HaloLabel[head].MinID > I->MinID)
-        {
-            HaloLabel[head].MinID = I->MinID;
-            HaloLabel[head].MinIDTask = I->MinIDTask;
-        }
-//        printf("unlocking %d by %d in ngbiter\n", other, omp_get_thread_num());
-        unlock_spinlock(head, spin);
+        /* update MinID.*/
+        __aint128 hminread;
+        do {
+            #pragma omp atomic read
```

Comment:
Is the atomic read needed? The atomic cas would probably block a partial read access?

If the atomic read is necessary for 128, then using a spinlock is definitely cleaner.

Reply:
I am not sure - I put it in only because I have been using it in the other CAS routines and it was there just for safety. If you think it is not needed I will remove it.

Reply:
I see. While for the 64 bit types atomic read compiles to a no-op, for the 128 bit types it is not, but has a call to GOMP_atomic_start (which probably locks a mutex). cmpxchg16b has a lock prefix which applies to the whole cacheline so tearing reads are not possible and the atomic read isn't needed. https://gcc.godbolt.org/z/3XU2Nk (compare the two reads).

I'm still worried though that if we don't get to use cmpxchg16b for some reason, losing the atomic read would cause half reads. __atomic_load()  compiles to another cmpxchg16 (!) with icc

Reply:
See the GOMP_atomic_start code: https://www.cs.rice.edu/~la5/doc/ligbomp-doc/db/d9c/critical_8c-source.html#l00125

I think I will remove the atomic read and if it is worth it performance wise, I will assert __atomic_always_lockfree(__int128) and maybe have a backup spinlock.